feat(service-worker): support multiple apps in one domain

sheikalthaf opened the pull request:
Current behavior of the service-worker is caching data of multiple apps in cache DB file in a domain.
Due to this behavior service-worker makes the apps to break when switching from one app to other app in same domain. This PR will solve this issue by suffixing the base-href with `ngsw` string to separate caches files of an apps in a domain

Fixes issue #21388

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #21388 
Currently service-worker doen't seperate the cache files based on base-href. Due to this caches files get clashes with multiple apps hosted in one domain


## What is the new behavior?
In this PR seperating the caches files by suffing the base-href with `ngsw` string like this `ngsw:<base-href>` to seperate the caches files of an app. This will avoid the storing of cache data in same file for multiple apps in one domain


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

sheikalthaf commented:
@gkalpak Please review this PR and let me know any changes required. Also need help in test writing

AndrewKushnir commented on packages/service-worker/worker/src/adapter.ts:
```diff
@@ -54,6 +55,17 @@ export class Adapter {
+   * in same domain with multiple apps
+   */
+  setBaseHref(baseHref: string) {
+    if (baseHref && ['/'].indexOf(baseHref) == -1) {
```
It looks like you can just compare base href with '/' string:
```
  if (baseHref && baseHref === '/') {
```

sheikalthaf replied:
@AndrewKushnir Hi,

ok now directly comparing with the '/'. Motivation is not to allow when base-href is not set that is '/'.
<pre><code>
if (baseHref && baseHref !== '/') {
</code></pre>
so that it work like current behavior

gkalpak commented:
@sheikalthaf, I see you are still pushing changes to the PR. Is this ready for review?

sheikalthaf commented:
@gkalpak yes it is now ready for review

gkalpak commented on packages/service-worker/worker/src/adapter.ts:
```diff
@@ -13,6 +13,7 @@
  * from the global scope.
  */
 export class Adapter {
+  ngsw = 'ngsw';
```
This should have a more descriptive name (e.g. `cachePrefix` or `cacheNamePrefix` or something).

sheikalthaf replied:
renamed to `cacheNamePrefix`

gkalpak commented on packages/service-worker/worker/src/adapter.ts:
```diff
@@ -54,6 +55,17 @@ export class Adapter {
+   * Suffixing the baseHref with `ngsw` string to avoid clash of cache files
+   * in same domain with multiple apps
+   */
+  setBaseHref(baseHref: string) {
```
IMO, this prefix should not be able to change at runtime. It should be determined during `Adapter` instantiation (as it remains constant throughout the lifetime of a SW instance).

gkalpak commented on packages/service-worker/worker/src/adapter.ts:
```diff
@@ -54,6 +55,17 @@ export class Adapter {
+   * in same domain with multiple apps
+   */
+  setBaseHref(baseHref: string) {
+    if (baseHref && baseHref !== '/') {
```
AFAICT, `baseHref` should never be empty (it would be at least `/`).
But even so, why this check? Why not use `/`? Is it to retain backwards compatibility with existing caches (thus avoiding re-fetching all assets from the server)?

gkalpak commented on packages/service-worker/worker/src/adapter.ts:
```diff
@@ -54,6 +55,17 @@ export class Adapter {
+   */
+  setBaseHref(baseHref: string) {
+    if (baseHref && baseHref !== '/') {
+      const str = baseHref.replace(/^\//, '').replace(/\/$/, '').replace(/\//, ':');
```
Why not use it as-is?

sheikalthaf replied:
This is why because '/' is not allowed in file name right?
if the base is something like this `/app/de/` then we have to replace them with ':' . so the result will be `app:de`.

gkalpak replied:
The cache name is not necessarily a file name. I didn't look at the spec, but Chrome seems to handle `/`s in cache names fine.

sheikalthaf replied:
ok thanks for the clarification. will remove the replace condition and use it as-is

gkalpak commented on packages/service-worker/worker/src/db-cache.ts:
```diff
@@ -23,16 +23,17 @@ export class CacheDatabase implements Database {
       this.tables.delete(name);
     }
-    return this.scope.caches.delete(`ngsw:db:${name}`);
+    return this.scope.caches.delete(`${this.adapter.ngsw}:db:${name}`);
```
In most cases (with the sole exception of the `control` DB), `CacheDatabase` methods (such as `delete`, `list`, `open`) are called with names that already contain `ngsw:<base-href>:`, so prepending it again will get a little too verbose. But it is safer. I'm on the fence on this one.

Maybe we could strip the leading `ngsw:<base-href>:` from the name (so that `baseHref` is only included in the final cache name once).

gkalpak commented on packages/service-worker/worker/src/driver.ts:
```diff
@@ -109,6 +109,12 @@ export class Driver implements Debuggable, UpdateSource {
+     * Extract baseHref from scope
+     * send the baseHref to adapter to suffix with ngsw string
+     */
+    const baseHref = new URL(scope.registration.scope).pathname;
```
IMO, this should be moved to `Adapter`'s instantiation.

sheikalthaf replied:
Moved to adapter

gkalpak commented on packages/service-worker/worker/src/driver.ts:
```diff
@@ -913,8 +919,8 @@ export class Driver implements Debuggable, UpdateSource {
-    const oldSwCacheNames =
-        cacheNames.filter(name => /^ngsw:(?:active|staged|manifest:.+)$/.test(name));
+    const regex = new RegExp(`^${this.adapter.ngsw}:(?:active|staged|manifest:.+)$`);
+    const oldSwCacheNames = cacheNames.filter(name => regex.test(name));
```
This change should be reverted. The `cleanupOldSwCaches()` method matches caches created by an old SW implementation.
In fact, it might be a good idea to also match (and delete) caches that don't have the pathname after `ngsw:`. AFAICT, `URL#pathname` will always start with a `/`, so I think it is safe to delete all caches matching `/^ngsw:(?!\/)/` (i.e. starting with `ngsw:` but not followed by `/`) - assuming we keep `baseHref`'s leading `/` in cache names.

sheikalthaf replied:
now changed the regex to above mentioned matching pattern. also adding baseHref without any changes to cache

gkalpak reviewed (changes requested):
The general direction this is going seems reasonable. Keep in mind that we also need tests that verify the new behavior.

sheikalthaf commented:
@gkalpak i have updated the code and spec but i don't know how to test this because the `hostUrl` is hardcode. Any guidance is helpfull.
Also i want any guide to stepup the angular development in windows machine

gkalpak commented on packages/service-worker/worker/testing/scope.ts:
```diff
@@ -254,6 +256,11 @@ export class SwTestHarness implements ServiceWorkerGlobalScope, Adapter, Context
     return promise;
   }
 
+  suffixBaseHref(hostUrl: string) {
```
Again, no need for a method here. We can interact with the property directly.

sheikalthaf replied:
removed this method

gkalpak reviewed (changes requested):
Thx for the changes, @sheikalthaf. I left some more comments.
As discussed "offline", the development workflow on Windows is a little problematic atm. You should be able to run the tests with `sh test.sh browserNoRouter` inside git bash (which comes bundled with git-for-windows).

sheikalthaf commented on packages/service-worker/worker/src/adapter.ts:
```diff
@@ -54,6 +61,15 @@ export class Adapter {
+   * Suffixing the baseHref with `ngsw` string to avoid clash of cache files
+   * in same domain with multiple apps
+   */
+  suffixBaseHref(hostUrl: string) {
```
changes commited

gkalpak reviewed (commented):
I've gone ahead and made some changes:
- Rebased on latest master.
- Squashed your commits into one.
- Added some more commits (including tests).

sheikalthaf commented on packages/service-worker/worker/src/adapter.ts:
```diff
@@ -13,6 +13,12 @@
  * from the global scope.
  */
 export class Adapter {
+  cacheNamePrefix: string;
```
Already it is marked as `readonly` in my branch.

sheikalthaf commented on packages/service-worker/worker/src/adapter.ts:
```diff
@@ -13,6 +13,12 @@
 export class Adapter {
+  cacheNamePrefix: string;
+  constructor(scope: ServiceWorkerGlobalScope) {
+    // Suffixing the baseHref with `ngsw` string to avoid clash of cache files
```
Meaningfull message 😊

sheikalthaf commented:
@gkalpak @IgorMinar @alxhub Waiting for this feature for a long time. Any news about this PR?

ASHFAQPATWARI commented:
@gkalpak @IgorMinar @alxhub  guys any update on this one. This is a blocker for us.

gkalpak commented:
Need one more from @alxhub or @IgorMinar.

IgorMinar reviewed (approved):
lgtm - thanks!!

IgorMinar commented:
can you please rebase this one more time and add merge-assistance label asking for cla override by a googler (because this PR contains commits from two authors and googler needs to okey it to prevent situations where people steal commits and we don't have CLA approval from the original author - obviously not the case here, but we still need an override)

IgorMinar commented:
and @sheikalthaf apologies for the delays in merging this in. we are working on improving the PR merging process to prevent these delays, but we still have a lot of work to get to the latency that would be more acceptable. We do appreciate your contribution! Let's get this rebased an in!

gkalpak committed:
refactor(service-worker): make second parameter to `Adapter#parseUrl()` optional

gkalpak committed:
test(service-worker): do not create testing artifacts on environments that do not support SW

The tests will not be run anyway, so the artifacts are never used and
there might be errors if creating the testing artifacts relies on APIs
that are not available in that environment (e.g. `URL`).

gkalpak committed:
test(service-worker): ensure `SwTestHarness#parseUrl()` behaves the same on browser and Node.js

gkalpak committed:
fixup! feat(service-worker): support multiple apps on different subpaths of a domain

sheikalthaf committed:
feat(service-worker): support multiple apps on different subpaths of a domain

Previously, it was not possible to have multiple apps (using
`@angular/service-worker`) on different subpaths of the same domain,
because each SW would overwrite the caches of the others (even though
their scope was different).

This commit fixes it by ensuring that the cache names created by the SW
are different for each scope.

Fixes #21388

gkalpak committed:
refactor(service-worker): use `Adapter#parseUrl()` for all URL parsing

This commit also ensures that the correct implementation is used on
environments that do not support `URL` (e.g. Node.js).

gkalpak committed:
test(service-worker): make it easy to use a different SW scope in tests

gkalpak committed:
test(service-worker): test support for multiple apps on different subpaths of a domain

gkalpak commented:
Rebased. Thx again for working on this, @sheikalthaf 👍

merge-assistance: Needs CLA override.

sheikalthaf commented:
@IgorMinar @gkalpak Thanks for your awesome support in helping contributors. On this PR special thanks to @gkalpak making this possible with his great help.
